ciao-cli/main.go

```diff
@@ -329,72 +326,104 @@ func limitToString(limit int) string {
 }
 
 func getCiaoEnvVariables() {
-	identity := os.Getenv(ciaoIdentityEnv)
 	controller := os.Getenv(ciaoControllerEnv)
-	username := os.Getenv(ciaoUsernameEnv)
-	password := os.Getenv(ciaoPasswordEnv)
-	tenant := os.Getenv(ciaoTenantNameEnv)
 	ca := os.Getenv(ciaoCACertFileEnv)
+	clientCert := os.Getenv(ciaoClientCertFileEnv)
 
 	infof("Ciao environment variables:\n")
-	infof("\t%s:%s\n", ciaoIdentityEnv, identity)
 	infof("\t%s:%s\n", ciaoControllerEnv, controller)
-	infof("\t%s:%s\n", ciaoUsernameEnv, username)
-	infof("\t%s:%s\n", ciaoPasswordEnv, password)
-	infof("\t%s:%s\n", ciaoTenantNameEnv, tenantName)
 	infof("\t%s:%s\n", ciaoCACertFileEnv, ca)
-
-	if identity != "" && *identityURL == "" {
-		*identityURL = identity
-	}
+	infof("\t%s:%s\n", ciaoClientCertFileEnv, clientCert)
 
 	if controller != "" && *controllerURL == "" {
 		*controllerURL = controller
 	}
 
-	if username != "" && *identityUser == "" {
-		*identityUser = username
+	if ca != "" && *caCertFile == "" {
+		*caCertFile = ca
 	}
 
-	if password != "" && *identityPassword == "" {
-		*identityPassword = password
+	if clientCert != "" && *clientCertFile == "" {
+		*clientCertFile = clientCert
 	}
+}
+
+func checkCompulsoryOptions() {
+	fatal := ""
 
-	if tenant != "" && *tenantName == "" {
-		*tenantName = tenant
+	if *clientCertFile == "" {
+		fatal += "Missing required client certificate file\n"
+	}
+	if *controllerURL == "" {
+		fatal += "Missing required Ciao controller URL\n"
 	}
 
-	if ca != "" && *caCertFile == "" {
-		*caCertFile = ca
+	if fatal != "" {
+		fatalf(fatal)
 	}
 }
 
-func checkCompulsoryOptions() {
-	fatal := ""
+func getTenantsFromCertFile(clientCertFile string) ([]string, error) {
+	var certBlock, p *pem.Block
+
+	data, err := ioutil.ReadFile(clientCertFile)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error loading client cert file")
+	}
 
-	if *identityURL == "" {
-		fatal += "Missing required identity URL\n"
+	for {
+		p, data = pem.Decode(data)
+		if p == nil {
+			break
+		}
+		if p.Type == "CERTIFICATE" {
+			certBlock = p
```

Comment:
Are we missing a break here, or are we looking for the last cert in the file, assuming there can be more than one?

Reply:
There will only be one PEM CERTIFICATE block in the file. If there are multiple x509 certificates (because there is a chain with an intermediate CA) they are still only encoded into a single PEM block.

However for safety i'll modify this to error out if p != nil when we try and make the assignment thus enforcing that behaviour.